db/sqlinterfaces.c

```diff
@@ -4991,33 +4990,39 @@ int dispatch_sql_query(struct sqlclntstate *clnt)
                 goto done;
             }
             int rc;
-            rc = pthread_cond_timedwait(&clnt->wait_cond, &clnt->wait_mutex,
-                                        &st);
+            rc = pthread_cond_timedwait(&clnt->wait_cond, &clnt->wait_mutex, &st);
             if (clnt->done) {
                 Pthread_mutex_unlock(&clnt->wait_mutex);
                 goto done;
             }
             if (rc == ETIMEDOUT) {
                 struct timespec diff;
-                TIMESPEC_SUB(st, last, diff);
-                if (diff.tv_sec >= clnt->heartbeat) {
-                    last = st;
-                    send_heartbeat(clnt);
-                    rc = fdb_heartbeats(clnt);
-                    if (rc)
-                        return -1;
-                }
                 if (clnt->query_timeout > 0 && !clnt->statement_timedout) {
                     TIMESPEC_SUB(st, first, diff);
                     if (diff.tv_sec >= clnt->query_timeout) {
-                        clnt->statement_timedout = 1;
-                        logmsg(LOGMSG_WARN, "%s:%d Query exceeds max allowed time %d.\n",
-                                __FILE__, __LINE__, clnt->query_timeout);
+                        if (lock_client_write_trylock(clnt) == 0) {
+                            clnt->statement_timedout = 1;
+                            logmsg(LOGMSG_WARN, "%s:%d Query exceeds max allowed time %d.\n", __FILE__, __LINE__, clnt->query_timeout);
```

Comment:
Add to longreqs instead?  It's a query that's taking longer than the user expects, sounds like a good candidate.  Doesn't have to be this PR.